Adding HA documentation (#2607)
openservicemesh/osm: docs/content/docs/HA.md

> ---

Author: I checked, In `index.md` it states:
> Use this directory to place proposals for features with detailed design explanation and any associated artifacts.

The "proposal" part made me back off from adding it there, as this tries to reflect current arch (not proposing anything new)

> ---

Author: ofc we can change it, just explaining why I didn't go for that

title: "HA Design considerations"
description: "Open Service Mesh HA Design considerations"
type: docs
---
# HA Design considerations

OSM is built with High Availability and Fault Tolerance in mind. The following sections will thoroughly document how these are tackled in OSM.

## High Availability and Fault Tolerance

High Availability and Fault Tolerance are implemented and ensured by several design decisions and external mechanisms in OSM, which will be documented in the following points:

### Statelessness
OSM does not own or have any state-dependent data that needs to be saved at runtime; with the controlled exceptions of:

> - Root certificate: when no root certificate is provided and Tresor is used, OSM will ensure atomicity and idempotence when creating the root certificate through Kubernetes secret API, even when multiple instances are spawned.

Author: +1 will clarify

> _Note: it is required and always assumed the root certificate is the same for all control plane components._

Author: I didn't call it a guarantee because there's room for misconfiguration in - for example - cert-manager. Since providing the root cert can sometimes depend on the user, I think two instances can still run with a separate root if the root CA is manually renewed and new instances spawn with a new root CA.

I think code can be put in place to enforce and guarantee in most circumstances that if the secret is changed, we reload it, but I believe code is not there right now (is it?).

> _Note: it is required and always assumed the root certificate is the same for all control plane components._

Reviewer: It should be a guarantee as long as one follows the prerequisites to provision to a certificate provider to use with OSM. If you misconfigure this, pods will not even connect to the controller.

> - Envoy Bootstrap Certificates (used by the proxies to authenticate against the control plane): these are created and issued through Kubernetes Secrets API at injector webhook handling time, and are ensured idempotence when being created.

Author: the bootrap config is still posted as a secret (cert inline and all) through k8s api (envoy_config.go @ L134), and loaded as a volume (patch.go @ L47)

I'm of the believe this is still partially true. Maybe I can reword it to reflect both what you pointed and the actual implementation.

Other than those exceptions, the rest of the configuration is built and fetched from Kubernetes.

> The domain state used to compute the traffic policies is entirely provided by the different providers (Kubernetes _et al_) and informers on the related objects OSM subscribes to.

Author: isn't it `go-client` ? :bat:

> Multiple OSMs running will subscribe to the same objects and should generate the same view over the cluster as long as Kubernetes providers remain eventually consistent when updating through the informers, rendering mesh topology and policy enforcement eventually consistent at Kubernetes API pace.

Reviewer: ```suggestion
Multiple `osm-controllers` running will subscribe to the same set of objects and will generate the identical configurations for the service mesh. Due to the nature of client-go Kubernetes informers being eventually consistent `osm-controller` guarantees policy enforcement to be eventually consistent.
```

Author: thanks, looks good.

<p align="center">

>   <img src="/docs/content/docs/images/ha/ha1.png" width="400" height="350"/>

Author: it's purely that. I decided to cut on few things for the sake of simplicity and clarity.

</p>

### Restartability: 
The previous stateless design considerations should ensure OSM control plane is fully restartable.

- A restarting instance will resynchronize all Kubernetes domain resources. Existing proxies will reconnect, and (assuming no changes occurred on the mesh topology or policy) the same configuration should be recomputed and pushed as a new version to the proxies.

### Horizontal scaling
OSM allows for horizontal scaling of its components (individually, for both the webhook/injector handling and the xDS control plane).

> - When OSM control plane is spawned with multiple replicaset, connecting proxies may be load-balanced and connected to any of the existing OSM instances running for the control plane.

Reviewer: ```suggestion
- When `osm-controller` is spawned with multiple replicaset, connecting proxies may be load-balanced and connected to any of the existing OSM instances running for the control plane.
```

> - Service certificates (used between proxies to TLS authenticate and communicate with each other) are short lived and kept only in runtime on control plane (though pushed as part of proxy xDS protocol when required).

Reviewer 2: They are cached in memory. All pods participating in a Service will get the same service cert - same CN.  Cert creation is expensive - so we cache it.  If the controller restarted and we lost the Service cert cache, we'll create a new one. The new one will have the same CN, same Issuing CA, but a different serial number, which is not an issue. This of course is for Tresor. If we were to use cert-manager.io, Hashi Vault, or Azure Key Vault - this would be different.

> - Service certificates (used between proxies to TLS authenticate and communicate with each other) are short lived and kept only in runtime on control plane (though pushed as part of proxy xDS protocol when required).

Reviewer: Certs are cached in memory for all certificate providers.

> - Service certificates (used between proxies to TLS authenticate and communicate with each other) are short lived and kept only in runtime on control plane (though pushed as part of proxy xDS protocol when required).

Author: > The new one will have the same CN, same Issuing CA, but a different serial number, which is not an issue
for the sake of discussion: It will eventually be an issue I believe. Restarting and re-issuing all certificates for all (connected) proxies and pushing all updates is going to take quite the toll at larger numbers.

>   Multiple OSM instances might create different yet valid service certificates for a single service. These different certificates will (1) have been signed by the same root, as multiple OSM instances must load the same root CA, and (2) will have the same Common Name (CN), which is be used to match against and authenticate when traffic is proxied between services.

Reviewer 3: ```suggestion
  Multiple `osm-controller` instances might create different yet valid service certificates for a single service. These different certificates will (1) have been signed by the same root, as multiple OSM instances must load the same root CA, and (2) will have the same Common Name (CN), which is be used to match against and authenticate when traffic is proxied between services.
```

<p align="center">
  <img src="/docs/content/docs/images/ha/ha2.png" width="400" height="350"/>
</p>

  In short, no matter what control plane a proxy connects to, a valid certificate with correct/proper CN will be pushed to it.

- Increasing horizontal scale will NOT redistribute established proxy connections to the control plane unless they are disconnected.
- Decreasing horizontal scale will make the disconnected proxies connect to instances that were not terminated by the downscale. New versions of the config should be computed and pushed upon establishing the connection anew.

<p align="center">
  <img src="/docs/content/docs/images/ha/ha3.png" width="400" height="350"/>
</p>

> - If the control plane is brought down entirely, running proxies should continue to operate in headless mode till they can reconnect to a running control plane.

Author: I thought headless is a rather well known term in the context of what we are working? -  oftentimes more related to networking (SDN cp/dp paradigm) but still applicable considering OSM is the control plane and the proxies act as DPs

> - If the control plane is brought down entirely, running proxies should continue to operate in headless mode till they can reconnect to a running control plane.

Author: I have the numbers. ping me anytime :)

The domain state used to compute the traffic policies is entirely provided by the different runtime providers (Kubernetes _et al_) and Kubernetes client-go informers on the related objects `osm-controller` subscribes to.

Multiple `osm-controller`s running will subscribe to the same set of objects and will generate the identical configurations for the service mesh. Due to the nature of client-go Kubernetes informers being eventually consistent `osm-controller` guarantees policy enforcement to be eventually consistent.

<p align="center">
  <img src="/docs/content/docs/images/ha/ha1.png" width="400" height="350"/>
</p>

### Restartability: 
The previous stateless design considerations should ensure OSM's control plane componenets are fully restartable.

- A restarting instance will resynchronize all Kubernetes domain resources. Existing proxies will reconnect, and (assuming no changes occurred on the mesh topology or policy) the same configuration should be recomputed and pushed as a new version to the proxies.

### Horizontal scaling
Components `osm-controller` and `osm-injector` allow for separate horizontal scaling, depending on load or availability requirements.

> - When an `osm-controller` is spawned with multiple replicas, connecting proxies may be load-balanced and connected to any of the existing OSM instances running for the control plane.

Reviewer 3: 👍
